[Merged by Bors] - add direction information to PeerInfo

blacktemplar opened the pull request:
## Issue Addressed

NA

## Proposed Changes

Adds a direction field to `PeerConnectionStatus` that can be accessed by calling `is_outgoing` which will return `true` iff the peer is connected and the first connection was an outgoing one.

realbigsean commented on beacon_node/eth2_libp2p/src/peer_manager/peer_info.rs:
```diff
@@ -309,8 +326,19 @@ impl PeerConnectionStatus {
     }
+
+    /// Returns true iff the peer is connected and its first connection was outgoing.
+    pub fn is_outgoing(&self) -> bool {
```
Might be nice adding `is_incoming` as well? Because it's not quite equal to `!is_outgoing`

realbigsean reviewed (commented):
Had one comment, looks good to me otherwise!

realbigsean commented:
I was staring at the API spec some more and noticed the example [here](https://ethereum.github.io/eth2.0-APIs/#/Node/getPeer) has a direction of `inbound` and a state of `disconnected`.  So the direction field must indicate the direction of the peer last time it was connected. So we'll have to track the direction outside of the `PeerConnectionStatus::Connected` enum variant or we'll lose the information when we disconnect. 

This seems a little weird to me though, because it also means that either: 
1. We always connect at some point to a peer we have knowledge of (is this the case? I don't know honestly) 
or: 
2. We only include peers we have connected to at some point in the `peer` endpoint responses, even if we know about others

Do you have any thoughts, @AgeManning ?

AgeManning commented:
Good questions. I made this issue to get some clarifications about the endpoint: https://github.com/ethereum/eth2.0-APIs/issues/97

blacktemplar commented:
Ok I changed the PR now so that the direction gets remembered also for disconnected/banned/dialing peers (if they had a connection before).

I also removed the `is_outgoing` method since I think its clearer to directly access the field `connection_direction` in the `PeerInfo`.

realbigsean commented:
nice! LGTM

blacktemplar commented:
Bors r+

AgeManning commented:
bors r+

AgeManning commented:
bors r+
